lib/sensu/server/filter.rb

```diff
@@ -165,28 +167,63 @@ def handle_severity?(handler, event)
         end
       end
 
-      # Ruby eval() a string containing an expression, within the
+      # Process a filter eval attribute, a Ruby `eval()` string
+      # containing an expression to be evaluated within the
+      # scope/context of a sandbox. This methods strips away the
+      # expression prefix, `eval:`, and substitues any dot notation
+      # tokens with the corresponding event data values. If there are
+      # unmatched tokens, this method will return `nil`.
+      #
+      # @event [Hash]
+      # @raw_eval_string [String]
+      # @return [String] processed eval string.
+      def process_eval_string(event, raw_eval_string)
+        eval_string = raw_eval_string.slice(5..-1)
+        eval_string, unmatched_tokens = substitute_tokens(eval_string, event)
+        if unmatched_tokens.empty?
+          eval_string
+        else
+          @logger.error("filter eval unmatched tokens", {
+            :raw_eval_string => raw_eval_string,
+            :unmatched_tokens => unmatched_tokens,
+            :event => event
+          })
+          nil
+        end
+      end
+
+      # Ruby `eval()` a string containing an expression, within the
       # scope/context of a sandbox. This method is for filter
       # attribute values starting with "eval:", with the Ruby
       # expression following the colon. A single variable is provided
       # to the expression, `value`, equal to the corresponding event
-      # attribute value. The expression is expected to return a
-      # boolean value.
+      # attribute value. Dot notation tokens in the expression, e.g.
+      # `:::mysql.user:::`, are substituted with the corresponding
+      # event data values prior to evaluation. The expression is
+      # expected to return a boolean value.
       #
+      # @param event [Hash]
       # @param raw_eval_string [String] containing the Ruby
       #   expression to be evaluated.
-      # @param value [Object] of the corresponding event attribute.
+      # @param raw_value [Object] of the corresponding event
+      #   attribute value.
       # @return [TrueClass, FalseClass]
-      def eval_attribute_value(raw_eval_string, value)
-        begin
-          eval_string = raw_eval_string.gsub(/\Aeval:(\s+)?/, "")
-          !!Sandbox.eval(eval_string, value)
-        rescue => error
-          @logger.error("filter attribute eval error", {
-            :raw_eval_string => raw_eval_string,
-            :value => value,
-            :error => error.to_s
-          })
+      def eval_attribute_value(event, raw_eval_string, raw_value)
+        eval_string = process_eval_string(event, raw_eval_string)
+        unless eval_string.nil?
+          begin
+            value = Marshal.load(Marshal.dump(raw_value))
```

Comment:
Marshal load()/dump() is used to deep dup any objects that are to be provided via `value`.